### PR TITLE
Big upgrades

### PR DESCRIPTION
**Pull request by joerunde**
#### Motivation

We need to update a whole bunch of things that will cause output differences, and we want to bundle them up together.

#### Modifications

Updates:
- pytorch
- flash attention
- autogptq
- cuda

#### Result

Slight differences in outputs for some text generation prompts on many models, but our quality tests indicate no major drop in result quality.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -265,8 +270,12 @@ RUN --mount=type=bind,from=flash-att-v2-cache,src=/usr/src/flash-attention-v2,ta
...
 
 # Copy over the auto-gptq wheel and install it
-RUN --mount=type=bind,from=auto-gptq-cache,src=/usr/src/auto-gptq-wheel,target=/usr/src/auto-gptq-wheel \
-    pip install /usr/src/auto-gptq-wheel/*.whl --no-cache-dir
+#RUN --mount=type=bind,from=auto-gptq-cache,src=/usr/src/auto-gptq-wheel,target=/usr/src/auto-gptq-wheel \
+#    pip install /usr/src/auto-gptq-wheel/*.whl --no-cache-dir
+
+# We only need to install a custom-built auto-gptq version if we need a pre-release
+# or are using a PyTorch nightly version
+RUN pip install auto-gptq=="${AUTO_GPTQ_VERSION}+cu118" --extra-index-url "https://huggingface.github.io/autogptq-index/whl/cu118/" --no-cache-dir
```
Comment (njhill) [inline Dockerfile:278]: If we are updating to CUDA 12 at the same time
```suggestion
RUN pip install auto-gptq=="${AUTO_GPTQ_VERSION}" --no-cache-dir
```

Comment (njhill) [reply Dockerfile:278]: @joerunde also it's possible we'll need to still build auto-gptq from source anyhow given the version differences of pytorch etc. See @cyang49's comment [here](https://github.com/IBM/text-generation-inference/pull/51#discussion_r1518460664).

Comment (joerunde) [reply Dockerfile]: Is it fine if that doesn't happen in this PR though?

**Comment by joerunde** [comment]
python package list from this image:
```
$ pip3 list | grep -iE '(flash|torch|auto|cuda)'
DEPRECATION: Loading egg at /opt/tgis/lib/python3.11/site-packages/custom_kernels-0.0.0-py3.11-linux-x86_64.egg is deprecated. pip 24.3 will enforce this behaviour change. A possible replacement is to use pip for package installation.. Discussion can be found at https://github.com/pypa/pip/issues/12330
auto_gptq                 0.7.1
flash-attn                2.5.6
nvidia-cuda-cupti-cu12    12.1.105
nvidia-cuda-nvrtc-cu12    12.1.105
nvidia-cuda-runtime-cu12  12.1.105
torch                     2.2.1+cu121
```

looks like the versions I expect- running performance and integration tests to make sure nothing is totally borked

File: Dockerfile (hunk 1 above)
Comment (njhill) [reply Dockerfile]: Yes if it works then no problem!

Comment (cyang49) [reply Dockerfile]: The error I saw might be caused by the pre-built package doesn't contain Marlin build flag. Which may mean that eventually we'll have to build from source if Marlin support is desired.

**Review by njhill — APPROVED** [review]
This is great, thanks @joerunde

**Pull request merged**